File: programmers/python/level2/kakao/2018/compression.py
import string
from collections import deque

def solution(msg):
    dic = {}
    chars = list(string.ascii_uppercase)
    for i, c in enumerate(chars) :
        dic[c] = i + 1
    last = 26

    msg = deque(msg)
    answer = []
    while(msg):
        ## 사전에 있는 가장 긴 문자열 찾기 
        now = ''
        temp = ''
        while True:
            now += msg.popleft()
            if not msg:
                break
            temp = now + msg[0]
            if not temp in dic :
                break
        answer.append(dic[now])
        last += 1
        if temp != '':
            dic[temp] = last
    return answer

# print(solution("KAKAO"))
# print(solution("TOBEORNOTTOBEORTOBEORNOT"))
# print(solution("ABABABABABABABAB"))

File: programmers/java/src/level2/kakao/kakao2018/Compression.java
package level2.kakao.kakao2018;

import java.util.*;
import java.util.stream.Collectors;

public class Compression {
    public static void main(String[] args) {
        Compression c = new Compression();
        System.out.println(c.solution("KAKAO"));
    }
    public List<Integer> solution(String msg) {
        Map<String, Integer> dic = new HashMap<>();
        for (int i = 1; i < 27; i++) {
             dic.put((char)(64+i) + "", i);
        }

        int last = 26;
        Queue<String> queue = Arrays.stream(msg.split("")).collect(Collectors.toCollection(LinkedList::new));
        List<Integer> answer = new ArrayList<>();

        while(!queue.isEmpty()) {
            String now = "";
            String temp = "";
            while(true) {
                now += queue.poll();
                if(queue.isEmpty()) {
                    break;
                }
                temp = now + queue.peek();
                if (! dic.containsKey(temp)) {
                    break;
                }
            }
            answer.add(dic.get(now));
            last++;
            if(temp != "") {
                dic.put(temp, last);
            }
        }
        return answer;
    }
}
